File: java/src/main/java/com/dokmatiq/docgen/builder/DocumentBuilder.java
package com.dokmatiq.docgen.builder;

import com.dokmatiq.docgen.client.DocumentsClient;
import com.dokmatiq.docgen.internal.FileUtils;
import com.dokmatiq.docgen.model.*;

import java.nio.file.Path;
import java.util.*;

/**
 * Fluent builder for constructing and executing document generation requests.
 *
 * <pre>{@code
 * byte[] pdf = dg.document()
 *     .html("<h1>Invoice {{nr}}</h1>")
 *     .template("invoice.odt")
 *     .field("nr", "2026-001")
 *     .watermark("DRAFT")
 *     .asPdf()
 *     .generate();
 * }</pre>
 */
public class DocumentBuilder {
    private DocumentsClient client;

    private String htmlContent;
    private String markdownContent;
    private String templateName;
    private String templateBase64;
    private Map<String, String> fields;
    private Map<String, String> bookmarks;
    private Map<String, String> markdownBookmarks;
    private List<ImageData> images;
    private List<QrCodeData> qrCodes;
    private List<BarcodeData> barcodes;
    private Map<String, TableData> tables;
    private PageSettings pageSettings;
    private Object watermark;
    private StationeryConfig stationery;
    private List<ContentArea> contentAreas;
    private InvoiceData invoiceData;
    private String password;
    private OutputFormat outputFormat;
    private String callbackUrl;
    private String callbackSecret;
    private Map<String, String> markdownStyles;

    /** @internal */
    public DocumentBuilder setClient(DocumentsClient client) {
        this.client = client;
        return this;
    }

    public DocumentBuilder html(String content) { this.htmlContent = content; return this; }
    public DocumentBuilder markdown(String content) { this.markdownContent = content; return this; }
    public DocumentBuilder template(String name) { this.templateName = name; return this; }

    public DocumentBuilder templateFile(Path file) {
        this.templateBase64 = FileUtils.toBase64(file);
        return this;
    }

    public DocumentBuilder field(String name, String value) {
        if (fields == null) fields = new LinkedHashMap<>();
        fields.put(name, value);
        return this;
    }

    public DocumentBuilder fields(Map<String, String> values) {
        if (fields == null) fields = new LinkedHashMap<>();
        fields.putAll(values);
        return this;
    }

    public DocumentBuilder bookmark(String name, String html) {
        if (bookmarks == null) bookmarks = new LinkedHashMap<>();
        bookmarks.put(name, html);
        return this;
    }

    public DocumentBuilder markdownBookmark(String name, String md) {
        if (markdownBookmarks == null) markdownBookmarks = new LinkedHashMap<>();
        markdownBookmarks.put(name, md);
        return this;
    }

    public DocumentBuilder image(String bookmarkName, Path file) {
        if (images == null) images = new ArrayList<>();
        images.add(new ImageData(bookmarkName, FileUtils.toBase64(file), null, null, null));
        return this;
    }

    public DocumentBuilder image(String bookmarkName, Path file, int width, int height) {
        if (images == null) images = new ArrayList<>();
        images.add(new ImageData(bookmarkName, FileUtils.toBase64(file), width, height, null));
        return this;
    }

    public DocumentBuilder qrCode(String bookmarkName, String content) {
        if (qrCodes == null) qrCodes = new ArrayList<>();
        qrCodes.add(new QrCodeData(bookmarkName, content, null, null, null));
        return this;
    }

    public DocumentBuilder barcode(String bookmarkName, String content, BarcodeFormat format) {
        if (barcodes == null) barcodes = new ArrayList<>();
        barcodes.add(new BarcodeData(bookmarkName, content, format, null, null));
        return this;
    }

    public DocumentBuilder table(String name, TableData data) {
        if (tables == null) tables = new LinkedHashMap<>();
        tables.put(name, data);
        return this;
    }

    public DocumentBuilder pageSettings(PageSettings settings) { this.pageSettings = settings; return this; }

    public DocumentBuilder watermark(String text) {
        this.watermark = text;
        return this;
    }

    public DocumentBuilder watermark(WatermarkConfig config) {
        this.watermark = config;
        return this;
    }

    public DocumentBuilder stationery(Path file) {
        this.stationery = new StationeryConfig(FileUtils.toBase64(file), null);
        return this;
    }

    public DocumentBuilder stationery(Path file, Path firstPage) {
        this.stationery = new StationeryConfig(FileUtils.toBase64(file), FileUtils.toBase64(firstPage));
        return this;
    }

    public DocumentBuilder contentArea(ContentArea area) {
        if (contentAreas == null) contentAreas = new ArrayList<>();
        contentAreas.add(area);
        return this;
    }

    public DocumentBuilder invoice(InvoiceData data) { this.invoiceData = data; return this; }
    public DocumentBuilder password(String pwd) { this.password = pwd; return this; }
    public DocumentBuilder outputFormat(OutputFormat format) { this.outputFormat = format; return this; }
    public DocumentBuilder asPdf() { this.outputFormat = OutputFormat.PDF; return this; }
    public DocumentBuilder asDocx() { this.outputFormat = OutputFormat.DOCX; return this; }
    public DocumentBuilder asOdt() { this.outputFormat = OutputFormat.ODT; return this; }

    public DocumentBuilder callback(String url, String secret) {
        this.callbackUrl = url;
        this.callbackSecret = secret;
        return this;
    }

    public DocumentBuilder markdownStyles(Map<String, String> styles) {
        this.markdownStyles = styles;
        return this;
    }

    /** Build the DocumentRequest without executing. */
    public DocumentRequest build() {
        return new DocumentRequest(
                htmlContent, markdownContent, templateName, templateBase64,
                fields, bookmarks, markdownBookmarks, images, qrCodes, barcodes,
                tables, pageSettings, watermark, stationery, contentAreas,
                invoiceData, password, outputFormat, callbackUrl, callbackSecret,
                markdownStyles
        );
    }

    /** Generate the document (requires attached client). */
    public byte[] generate() {
        if (client == null) {
            throw new IllegalStateException(
                    "No client attached. Use dg.document() or pass the request to dg.documents().generate().");
        }
        return client.generate(build());
    }
}

File: typescript/src/builders/document-builder.ts
import type { DocumentsClient } from "../clients/documents.js";
import type {
  ContentArea,
  StationeryConfig,
  WatermarkConfig,
} from "../models/content.js";
import type {
  DocumentRequest,
  MarkdownStyles,
} from "../models/document.js";
import type { OutputFormat, BarcodeFormat } from "../models/enums.js";
import type { InvoiceData } from "../models/invoice.js";
import type {
  BarcodeData,
  ImageData,
  QrCodeData,
  TableData,
} from "../models/media.js";
import type { PageSettings } from "../models/page.js";
import { toBase64, type FileInput } from "../files.js";

/**
 * Fluent builder for constructing and executing document generation requests.
 *
 * @example
 * ```ts
 * const pdf = await dg.document()
 *   .html('<h1>Invoice {{nr}}</h1>')
 *   .template('invoice.odt')
 *   .field('nr', '2026-001')
 *   .watermark('DRAFT')
 *   .asPdf()
 *   .generate();
 * ```
 */
export class DocumentBuilder {
  private readonly request: DocumentRequest = {};
  private client?: DocumentsClient;

  /** @internal Attach a client for direct generation. */
  _setClient(client: DocumentsClient): this {
    this.client = client;
    return this;
  }

  /** Set the HTML content. */
  html(content: string): this {
    this.request.htmlContent = content;
    return this;
  }

  /** Set the Markdown content. */
  markdown(content: string): this {
    this.request.markdownContent = content;
    return this;
  }

  /** Set the template by name (must be pre-uploaded). */
  template(name: string): this {
    this.request.templateName = name;
    return this;
  }

  /** Set the template from a file (uploaded inline as base64). */
  templateFile(file: FileInput): this {
    this.request.templateBase64 = toBase64(file);
    return this;
  }

  /** Set a single template field. */
  field(name: string, value: string): this {
    this.request.fields ??= {};
    this.request.fields[name] = value;
    return this;
  }

  /** Set multiple template fields at once. */
  fields(values: Record<string, string>): this {
    this.request.fields = { ...this.request.fields, ...values };
    return this;
  }

  /** Set a bookmark replacement. */
  bookmark(name: string, html: string): this {
    this.request.bookmarks ??= {};
    this.request.bookmarks[name] = html;
    return this;
  }

  /** Set a markdown bookmark replacement. */
  markdownBookmark(name: string, markdown: string): this {
    this.request.markdownBookmarks ??= {};
    this.request.markdownBookmarks[name] = markdown;
    return this;
  }

  /** Add an image at a bookmark. */
  image(bookmarkName: string, file: FileInput, options?: Omit<ImageData, "bookmarkName" | "base64">): this {
    this.request.images ??= [];
    this.request.images.push({
      bookmarkName,
      base64: toBase64(file),
      ...options,
    });
    return this;
  }

  /** Add a QR code at a bookmark. */
  qrCode(bookmarkName: string, content: string, options?: Omit<QrCodeData, "bookmarkName" | "content">): this {
    this.request.qrCodes ??= [];
    this.request.qrCodes.push({ bookmarkName, content, ...options });
    return this;
  }

  /** Add a barcode at a bookmark. */
  barcode(
    bookmarkName: string,
    content: string,
    format?: BarcodeFormat,
    options?: Omit<BarcodeData, "bookmarkName" | "content" | "format">,
  ): this {
    this.request.barcodes ??= [];
    this.request.barcodes.push({
      bookmarkName,
      content,
      format,
      ...options,
    });
    return this;
  }

  /** Add table data at a named position. */
  table(name: string, data: TableData): this {
    this.request.tables ??= {};
    this.request.tables[name] = data;
    return this;
  }

  /** Set page layout settings. */
  pageSettings(settings: PageSettings): this {
    this.request.pageSettings = settings;
    return this;
  }

  /** Add a diagonal watermark overlay. */
  watermark(text: string, options?: Omit<WatermarkConfig, "text">): this {
    if (options) {
      this.request.watermark = { text, ...options };
    } else {
      this.request.watermark = text;
    }
    return this;
  }

  /** Set stationery (letterhead) PDF background. */
  stationery(file: FileInput, firstPageFile?: FileInput): this {
    const config: StationeryConfig = { pdfBase64: toBase64(file) };
    if (firstPageFile) {
      config.firstPagePdfBase64 = toBase64(firstPageFile);
    }
    this.request.stationery = config;
    return this;
  }

  /** Add an absolutely positioned content area. */
  contentArea(area: ContentArea): this {
    this.request.contentAreas ??= [];
    this.request.contentAreas.push(area);
    return this;
  }

  /** Attach structured invoice data (ZUGFeRD/XRechnung). */
  invoice(data: InvoiceData): this {
    this.request.invoiceData = data;
    return this;
  }

  /** Set a password for the output document. */
  password(pwd: string): this {
    this.request.password = pwd;
    return this;
  }

  /** Set the output format. */
  outputFormat(format: OutputFormat): this {
    this.request.outputFormat = format;
    return this;
  }

  /** Shorthand: set output to PDF. */
  asPdf(): this {
    this.request.outputFormat = "PDF";
    return this;
  }

  /** Shorthand: set output to DOCX. */
  asDocx(): this {
    this.request.outputFormat = "DOCX";
    return this;
  }

  /** Shorthand: set output to ODT. */
  asOdt(): this {
    this.request.outputFormat = "ODT";
    return this;
  }

  /** Set the async callback URL. */
  callback(url: string, secret?: string): this {
    this.request.callbackUrl = url;
    if (secret) {
      this.request.callbackSecret = secret;
    }
    return this;
  }

  /** Set Markdown rendering style overrides. */
  markdownStyles(styles: MarkdownStyles): this {
    this.request.markdownStyles = styles;
    return this;
  }

  /** Build the DocumentRequest object without executing. */
  build(): DocumentRequest {
    return { ...this.request };
  }

  /** Generate the document (requires attached client). */
  async generate(): Promise<Buffer> {
    if (!this.client) {
      throw new Error(
        "No client attached. Use dg.document() or pass the request to dg.documents.generate().",
      );
    }
    return this.client.generate(this.build());
  }
}
